### DISCUSSION TITLE
How to make a MainWindow's button interact with a MainViewModel's Property?

### DISCUSSION THREAD

**Discussion by Nemonek** [AvaloniaUI/Avalonia#14426] [Q&A] [answered]
Hi, I'm a student currently trying to learn avalonia and the MVVM pattern: I'm creating an application and I stumbled into a problem.
The problem is that I have a MainWindow.axaml with inside some buttons positioned next to a ViewModel: I need a way to make those buttons interact with data inside the MainViewModel.
![immagine](https://github.com/AvaloniaUI/Avalonia/assets/127589992/2631f502-592d-4d3f-8920-82a63dffb073)
The button on the left side are inside the MainWindow.axaml file, the template of the items you see are defined inside a ListBox in the MainView.axaml file, and the ObservableCollection wich contains the data binded to the ListBox are inside the MainViewModel.cs file: I need some way to make the button in the MainWindow interact with the MainViewModel's ObservableCollection. I cannot transfer the ListBox to the MainWindow and delete the MainView cause I need to swap the View visualized when a click occurs on the "Abilities" button.

The code for the MainWindow:
```axaml
<Window.Styles>
	<StyleInclude Source="/LeftMenuButtonStyles.axaml"/>
</Window.Styles>


<Grid RowDefinitions="50, *, 50">
	<Grid.ColumnDefinitions>
		<ColumnDefinition MaxWidth="300" Width="*"></ColumnDefinition>
		<ColumnDefinition Width="4.5*"></ColumnDefinition>
	</Grid.ColumnDefinitions>
	<TextBlock HorizontalAlignment="Center" VerticalAlignment="Center" FontSize="20" FontWeight="Bold" Grid.Column="0" Grid.ColumnSpan="2" Grid.Row="0" Text="{Binding Character.Name}" />
	<Border Grid.Row="1" Grid.Column="0" Padding="0" Margin="0" BorderBrush="Gray" BorderThickness="0, 0, 1, 0"></Border>
	<StackPanel Grid.Row="1" Grid.Column="0">

		<Button Grid.Column="0" Classes="MenuPrincipale">
			<DockPanel>
				<TextBlock>Abilities</TextBlock>
				<PathIcon Height="15" VerticalAlignment="Center" HorizontalAlignment="Right" Data="{StaticResource chevron_right_regular}"/>
			</DockPanel>
		</Button>
		<Button Grid.Column="0" Command="{Binding ShowAllAvailable}" Classes="MenuPrincipale">
			<DockPanel>
				<TextBlock>All</TextBlock>
				<PathIcon Height="15" VerticalAlignment="Center" HorizontalAlignment="Right" Data="{StaticResource chevron_right_regular}"/>
			</DockPanel>
		</Button>
		<Button Command="{Binding FilterByReady}" Classes="MenuPrincipale">
			<DockPanel>
				<TextBlock>Ready</TextBlock>
				<PathIcon Height="15" VerticalAlignment="Center" HorizontalAlignment="Right" Data="{StaticResource chevron_right_regular}"/>
			</DockPanel>
		</Button>
		<Button Command="{Binding FilterByRitual}" Classes="MenuPrincipale">
			<DockPanel>
				<TextBlock>Ritual</TextBlock>
				<PathIcon Height="15" VerticalAlignment="Center" HorizontalAlignment="Right" Data="{StaticResource chevron_right_regular}"/>
			</DockPanel>
		</Button>
		<Button Command="{Binding FilterByNotReady}" Classes="MenuPrincipale">
			<DockPanel>
				<TextBlock>Not ready</TextBlock>
				<PathIcon Height="15" VerticalAlignment="Center" HorizontalAlignment="Right" Data="{StaticResource chevron_right_regular}"/>
			</DockPanel>
		</Button>
		<Button Command="{Binding ShowSpellInfoPopup}" Classes="MenuPrincipale">
			<DockPanel>
				<TextBlock>PROVA</TextBlock>
				<PathIcon Height="15" VerticalAlignment="Center" HorizontalAlignment="Right" Data="{StaticResource chevron_right_regular}"/>
			</DockPanel>
		</Button>
	</StackPanel>
	<views:MainView Grid.Row="1" Grid.Column="1" />
	
</Grid>
```
As you can see in this code the buttons are binded to a command: those command need to interact with the MainViewModel's data.
Right now I setted the MainWindow DataContext to the MainViewModel instance, but since I need to change the View displayed in the MainWindow and have different buttons that interacts with data's in different views I need to have only one DataContext for all the commands I need for those buttons.
Here's the code of the MainViewModel.cs: you can see that the command that right now are called by the button in the MainWindow operates directly on the collection binded to the ListBox in the MainView.axaml file
```C#
public class MainViewModel : ViewModelBase, INotifyPropertyChanged
{
    private Wizard _character;
    public Wizard Character => this._character;

    private ObservableCollection<Spell> _itemSourceForListBox;
    public ObservableCollection<Spell> ItemSourceForListBox { 
        get { return this._itemSourceForListBox; } 
        private set {
            this._itemSourceForListBox = value;
            OnPropertyChanged(nameof(this.ItemSourceForListBox));
        } 
    }

    public MainViewModel()
    {
        this._character = new("test1", "School");
        this.ItemSourceForListBox = this._character.AvailableSpells;
        ShowAllAvailable = ReactiveCommand.Create(() => this.ItemSourceForListBox = this._character.AvailableSpells);
        FilterByReady = ReactiveCommand.Create(() => this.ItemSourceForListBox = new(this._character.AvailableSpells.Where(t => t.IsReady)));
        FilterByRitual = ReactiveCommand.Create(() => this.ItemSourceForListBox = new(this._character.AvailableSpells.Where(t => t.IsRitual)));
        FilterByNotReady = ReactiveCommand.Create(() => this.ItemSourceForListBox = new(this._character.AvailableSpells.Where(t => !t.IsReady)));

        ShowSpellInfoPopup = ReactiveCommand.Create((Spell s) => new SpellInfo(s).Show()); // warning
    }

    // Commands
    public ICommand ShowAllAvailable { get; }
    public ICommand FilterByReady { get; }
    public ICommand FilterByRitual { get; }
    public ICommand FilterByNotReady { get; }

    public ICommand ShowSpellInfoPopup { get; } // warning

    // INotifyPropertyChanged implementation
    public event PropertyChangedEventHandler PropertyChanged;
    protected virtual void OnPropertyChanged(string propertyName)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}

```

Also, another question, why when I created the project I got provided with a MainWindow.axaml, a MainView.axaml and a MainViewModel.cs but not a MainWindowModel.cs? ( The data context of the MainWindow is automatically setted in the App.axaml.cs file to a MainViewModel instance )

If I missed something or something's unclear feel free to tell me and I'll provide as soon as possible to fix it.
Every advice is welcome!
Thank you!

**Comment by timunie** [comment]
Maybe check out https://github.com/AvaloniaUI/Avalonia.Samples

**Reply by Nemonek** [reply]
I looked at this [https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/MVVM/BasicMvvmSample](url), and the first approach seems a valid solution to my problem, but if I want to use this method I'd have another problem. The object I'm working on ( I'll call it wizard ) in this case needs to be in the MainWindowViewModel, but the MainViewModel and another view I'm developing needs access to it. I read that passing it through the constructor would be a MVVM violation: is this true and I need another approach? Or can I just pass it around thi way?

**Reply by thevortexcloud** [reply]
VMs can reference each other without violating MVVM. Views can reference any VM they like. VMs can't reference views. To quote Microsoft:

> In addition to understanding the responsibilities of each component, it's also important to understand how they interact. At a high level, the view "knows about" the view model, and the view model "knows about" the model, but the model is unaware of the view model, and the view model is unaware of the view. Therefore, the view model isolates the view from the model, and allows the model to evolve independently of the view.


> The view model is also responsible for coordinating the view's interactions with any model classes that are required. There's typically a one-to-many relationship between the view model and the model classes. The view model might choose to expose model classes directly to the view so that controls in the view can data bind directly to them. In this case, the model classes will need to be designed to support data binding and change notification events.
>
> Each view model provides data from a model in a form that the view can easily consume. To accomplish this, the view model sometimes performs data conversion. Placing this data conversion in the view model is a good idea because it provides properties that the view can bind to. For example, the view model might combine the values of two properties to make it easier to display by the view.

The fact VMs can reference other VMs is a big part of how the view locator that ships with Avalonia works. It can just create a lot of coupling if you do it too much. But sometimes there is no better way. In any case, your VMs in general should be fairly "dumb". Which is to say all your application logic should be elsewhere. They are just responsible for marshalling the data around and getting it ready for presentation.

**Reply by Nemonek** [reply]
Ok, I can work with that, and I'll probably do so; my problem is different ( I probably haven't explaine well ): I have an object which contains all the data and informations I need to display: this object needs to be referenced from the MainWindowModel, in which it'll be created, the MainViewModel, and the SecondViewModel: should I pass this object to the other VMs using their constructors?

**Reply by thevortexcloud** [reply]
It sounds kind of like you want DI? Which is also something I do myself _a lot._ Constructor injection is a big part of that. But you don't really need to use DI to pass common objects/services around. It just makes much easier. If you do it via the DI route, it's best practice to abstract as much as you can into interfaces (even if those interfaces only have a single implementation) to prevent direct coupling and make unit testing significantly easier to do. 

Anyway, long as you are not referencing any part of your views in your VM, you are fine. The VM can contain references to pretty much anything else. EG I regularly inject a reference to a configuration service, a database service, etc into my VMs.

That said, if you are sharing a non immutable object you do have to worry about what each of the VMs is doing with it as they may end up changing values while you are in the middle of using it elsewhere. Which is why DI services are typically designed to do one very specific thing and are typically stateless/contain minimal state.

**Reply by Nemonek** [reply]
I honestly never looked deep into DI, I just need a reference to the object in every model ( maybe when expanding the application I'm working on I'll need DI, but for now I'll focus on implementing the first part ).
Thank you!

**Reply by thevortexcloud** [reply]
> I just need a reference to the object in every model

Honestly you are probably better off just using a static value at that point if you are not going to bother with DI. Just keep in mind it may may writing unit tests harder.

**Reply by Nemonek** [reply]
> Just keep in mind it may may writing unit tests harder.
Do you mean "it may make writing unit test harder"?

Last question, then I won't have any more problem: I have 5 buttons, the first needs to change the view to the 2 one, the other 4 needs to make some operations in the date displayed in the first view, and, if necessary, change the view from the second to the first ( for which they're operating ). To make the view change based on the button clicked I was thinking about implementing something like this [https://github.com/SingletonSean/wpf-tutorials/tree/master/SwitchingViewsMVVM](url). 
And each of the 4 buttons all set the view to the first, an then make the operation on the data.
Would this be a good approach? 
Also, when changing the view, how can I set the Data context of each view while passing to them the object?

**Reply by thevortexcloud** [reply]
> Do you mean "it may make writing unit test harder"?

Whoops. Yes, that is what I meant. It's because you can't mock concrete types and you also can't override the implementation used with static types easily.

> [https://github.com/SingletonSean/wpf-tutorials/tree/master/SwitchingViewsMVVM](https://github.com/AvaloniaUI/Avalonia/discussions/url)

I am not sure why your link is pointing to somewhere else entirely than what the actual text is.

In any case: This looks like just a very simple hardcoded routing pattern. There are better and more flexible ways to to do this. If you are using ReactveUI for example, you can use the built in router.

https://docs.avaloniaui.net/docs/concepts/reactiveui/routing

Or you can write your own. I wrote a prototype router of my own awhile back that works using "URLs" kind of like a website link. You just provide the full path and it will automagically make it appear when the button is pressed based on what ever routes it has discovered. The advantage of that approach is you can keep it very generic and not worry about where your VMs come from (in my case they came from a DI container). It also stored page history so it was possible to navigate backwards. So the the XAML of my top level view looked like this:

```xaml
    <Grid ColumnDefinitions="auto *">
        <StackPanel Orientation="Vertical">
            <Button Content="Main" Command="{CompiledBinding GoToPageCommand}" CommandParameter="ViewModels/First" />
            <Button Content="Test" Command="{CompiledBinding GoToPageCommand}" CommandParameter="ViewModels/Second" />
            <Button Content="Back" Command="{CompiledBinding GoBackCommand}" />
        </StackPanel>
        <ContentControl Grid.Column="1" Content="{CompiledBinding CurrentPage}" />
    </Grid>
```

The downside with that type of router is you are dealing with strings, instead of strongly typed references. 

There are many other routers available though, some more simple than others.

That said, if you don't need a full router: that approach will be fine. I personally though would not make a dedicated command for it. It seems kind of weird to me. I would make it so each page has its own command, and then have some public view model property and bind it to a content control, then update it to what ever view model it needs to be set to.

>  I set the Data context of each view while passing to them the object?

You would not. You would pass your VM to some sort of content control (like I have above) and then let the template system figure out how to display it. EG use a view locator, or a template selector.

**Reply by Nemonek** [reply]
Hey, sorry for the late reply, I was busy all yesterday afternoon and today's morning, and it also took me some time before understanding what a ViewLocator does. So, like I said, I understood what's the goal of having a ViewLocator, and I used it ( I also switched my application to the MVVM template, so it took even more time ). Right now I have a button binded to a command like this
```C#
// This is what's in the constructor, there is also a public property of type ICommand that makes this command accessible
  this.ChangeViewToSpellsView = ReactiveCommand.Create(() =>
  {
      this.CurrentView = this._spellsView;
  });
```
and this command sets the CurrentView to an SpellsViewModel instance
```C#
  public ViewModelBase CurrentView { 
      get => this._currentView;
      private set
      {
          this._currentView = value;
          OnPropertyChanged(nameof(CurrentView));
      }

  }
  private SpellsViewModel _spellsView;
```
And in the MainWindow.axaml I have this
```C#
<ContentControl Grid.Row="1" Grid.Column="1" Content="{Binding CurrentView}"/>
```
When the button gets pressed the CurrentView property gets setted to the instance of SpellsViewModel, and this seems to work, the only problem is that it's a little slow, not too much, but when I click to change the view you can see that something is happening behind everything: is there a way to remove this little latency?

The last problem is that I have those 4 buttons:
![immagine](https://github.com/AvaloniaUI/Avalonia/assets/127589992/ad1c9d8d-6153-481a-96f1-5f6a4cde05af)
When I click on one of those I need the view to automatically change to the second one and also make some changes to the data inside the DataContext ( I'll call it SecondView ): would be a problem creating 4 different ICommand that does this:
```C#
  this.ChangeViewToSecondView = ReactiveCommand.Create(() =>
  {
      this.CurrentView = this._SecondView;
      // Some operation
  });
```
All the ViewModels have a copy of the reference to the main object I'm operating on and the CurrentView object in the MainWindowViewModel
Also, this group of buttons I'm talking about is inside a stack panel with DataContext a SecondViewModel instance, the first button, that switches to the SpellsView have for DataContext the MainViewModel: is this a correct practice? Or it's just bad?

Sorry again for the late reply!

**Reply by timunie** [reply]
looks okay to me. Performance should be judged at release mode only. Also if you load some data during navigation, make sure this is going to be async if possible

**Reply by Nemonek** [reply]
Thank you! For the data I'm trying to create a local EF database with sqlite ( I'm a student, so that will take some time ), but thank for the advice!
I'll mark this question as answered!
Thanks to all of you!

**Reply by thevortexcloud** [reply, marked as answer]
> Also, this group of buttons I'm talking about is inside a stack panel with DataContext a SecondViewModel instance, the first button, that switches to the SpellsView have for DataContext the MainViewModel: is this a correct practice? Or it's just bad?

I am not sure I understand. Can you elaborate/reword this?

As for your delay: You also don't seem to be using async which might not be helping. Do note, that for weird reasons some async SQL operations may be significantly slower than synchronous ones. But I can't remember if the SQLite provider has that problem.

https://github.com/dotnet/SqlClient/issues/593

Also, .NET often lazy loads assemblies. If you are relying on code in another assembly there can sometimes be a bit of stuttering the first time you use code from it. EF for example is a fairly large library (and also kind of weird to use in a desktop app, it's really meant to be used on web servers where the initial start up time does not matter that much). The core package is ~6mb, and that's not including all the other stuff it needs for SQLite. Dapper (a much simpler ORM) by comparison is only ~600kb. But it's entirely possible you are running into a different problem here, it's hard to say without seeing more of your code. 

Personally though, I would just use Dapper. I have used it with various Avalonia projects in the past and it works great. The main downside is it does not abstract the SQL for you, so you will need to write your own queries.

However if you are using an IDE you can also try profiling your app to see what exactly is so slow. I have found some fairly inefficient view model update code in the past by doing that.

**Reply by Nemonek** [reply]
> I am not sure I understand. Can you elaborate/reword this?

I solved that problem, I'm now using the MessageBus class to send messages between the view, and I use the viewlocator to get the corresponding view based on the model's instance. <br>
Anyway, when I wrote that I had two group of buttons in MainWindow.axaml: each with a different DataContext ( more precisely the first had a MainViewModel instance, and the secon group had a SecondViewModel instance ( those names are probably different, but I don't remember precisely cause I changed nearly everything in the program ) ): the question were if having something like this is good or bad.
<br>
As for the database, it's not implemented yet: I tried on a copy of the project, and I made it work ( I think ), but in the actual program is not there yet. As for the assemblies, I'll look into Dapper, cause I just need a little local database to work with. <br>
To conclude, the problem of the speed seems to go away after the first change of view, maybe it's just somwthing that needs to be initialized. I'll try to profile my app anyway, who knows, maybe, or probably, I'll find some problem.

Since all my problems seems to be solved, this evening I'm going to close this question ( I forgot the other day ).
Thank you all for the help!

**Reply by thevortexcloud** [reply]
> I'm now using the MessageBus class to send messages between the view

If you are using Reactive UI you can also use interactions.

https://www.reactiveui.net/docs/handbook/interactions/

>  I'll look into Dapper, cause I just need a little local database to work with. 

It works pretty well for in my experience. Far nicer than raw ADO code. Which is fairly painful to use. EF is fairly complicated and can take a long time to learn properly, even for an experienced .NET dev.

> To conclude, the problem of the speed seems to go away after the first change of view, maybe it's just somwthing that needs to be initialized. 

Sounds like an assembly loading/caching problem to me.

**Comment by thevortexcloud** [comment]
> why when I created the project I got provided with a MainWindow.axaml, a MainView.axaml and a MainViewModel.cs but not a MainWindowModel.cs?

Did you pick the wrong template? Avalonia ships with an MVVM template and a non MVVM template. MVVM is completely optional, although not using it can be fairly painful.

**Reply by Nemonek** [reply]
Ah, I selected Avalonia C# Project, so yes, I think I picked the wrong template; I'll provide to fix; thank you!
